### PR TITLE
optimizer/plan: check join path subquery when attach condition.

### PR DESCRIPTION
**Pull request by coocood**
Fixes https://github.com/pingcap/tidb/issues/1031

### CODE REVIEW

File: session_test.go [hunk 1]
```diff
@@ -1647,6 +1647,18 @@ func (s *testSessionSuite) TestIgnoreForeignKey(c *C) {
 	mustExecSQL(c, se, sqlText)
 }
 
+func (s *testSessionSuite) TestJoinSublquery(c *C) {
```
Comment (qiuyesuifeng) [inline session_test.go]: Why not using `TestJoinSubquery`?

**Comment by qiuyesuifeng** [comment]
LGTM

File: session_test.go [hunk 2]
```diff
@@ -1647,6 +1647,18 @@ func (s *testSessionSuite) TestIgnoreForeignKey(c *C) {
...
 
+func (s *testSessionSuite) TestJoinSubquery(c *C) {
+	store := newStore(c, s.dbName)
+	se := newSession(c, store, s.dbName)
+	mustExecSQL(c, se, "CREATE TABLE table1 (id INTEGER key AUTO_INCREMENT, data VARCHAR(30))")
+	mustExecSQL(c, se, "CREATE TABLE table2 (id INTEGER key AUTO_INCREMENT, data VARCHAR(30), t1id INTEGER)")
+	sqlTxt := `SELECT table1.id AS table1_id, table1.data AS table1_data FROM
+	table1 INNER JOIN (
+		SELECT table2.id AS id, table2.data AS ata, table2.t1id AS t1id FROM table2
+	) AS anon_1 ON table1.id = anon_1.t1id;`
```
Comment (zimulala) [inline session_test.go]: ata -> data?

File: session_test.go [hunk 3]
```diff
@@ -1647,6 +1647,18 @@ func (s *testSessionSuite) TestIgnoreForeignKey(c *C) {
...
 }
 
+func (s *testSessionSuite) TestJoinSubquery(c *C) {
+	store := newStore(c, s.dbName)
+	se := newSession(c, store, s.dbName)
+	mustExecSQL(c, se, "CREATE TABLE table1 (id INTEGER key AUTO_INCREMENT, data VARCHAR(30))")
+	mustExecSQL(c, se, "CREATE TABLE table2 (id INTEGER key AUTO_INCREMENT, data VARCHAR(30), t1id INTEGER)")
+	sqlTxt := `SELECT table1.id AS table1_id, table1.data AS table1_data FROM
+	table1 INNER JOIN (
+		SELECT table2.id AS id, table2.data AS ata, table2.t1id AS t1id FROM table2
```
Comment (zimulala) [inline session_test.go]: ata -> data?

Comment (coocood) [inline session_test.go]: This is ported from ORM code.

Comment (ngaut) [inline session_test.go]: Looks like it's a typo. We should correct it.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**